### PR TITLE
4o migrate backend

### PR DESCRIPTION
**Pull request by rishabhy**

### CODE REVIEW

File: src/openai_public.rs [hunk 1]
```diff
@@ -25,92 +25,117 @@ const IM_SEP: &str = "<|im_sep|>";
...
+    let special_tokens: HashMap<String, usize> =
+      [(ENDOFTEXT.to_string(), 199999), (ENDOFPROMPT.to_string(), 200018)]
+        .iter()
+        .cloned()
+        .collect();
+
+    const PAT_STR: &str = concat!(
+      r"[^\r\n\p{L}\p{N}]?[\p{Lu}\p{Lt}\p{Lm}\p{Lo}\p{M}]*[\p{Ll}\p{Lm}\p{Lo}\p{M}]+(?:'s|'t|'re|'ve|'m|'ll|'d)?|",
+      r"[^\r\n\p{L}\p{N}]?[\p{Lu}\p{Lt}\p{Lm}\p{Lo}\p{M}]+[\p{Ll}\p{Lm}\p{Lo}\p{M}]*(?:'s|'t|'re|'ve|'m|'ll|'d)?|",
+      r"\p{N}{1,3}|",
```
Comment (arvid220u) [inline src/openai_public.rs]: why is the pat str different?

i believe python just passes the raw pat string into rust.

Comment (arvid220u) [reply src/openai_public.rs]: also, let's use the same syntax they use in python for less confusion (i.e. writing a list [] with a .join, unless there's a reason that that doesn't work)

**Review by arvid220u — CHANGES_REQUESTED** [review]
let's add one or two tests to tests.rs

just make some new txt files and follow the existing pattern. ideally, let's include a string that would trigger the `i` behavior (i believe the current regex would have an incorrect token count for `I'VE, THEY'RE, it's, I'd, Would've, would'VE WOULD'VE` but i could be wrong)

for counting actual tokens, i'd rely on sending an actual request to the openai api and seeing the result tokens (i wouldn't trust the tiktoken website, for example)

**Comment by arvid220u** [comment]
also, let's make a o200k_im tokenizer too! in our backend, we don't use cl100k_base, but instead use cl100k_im, which has the im special tokens (see other pr for how to find those tokens)

after adding the im special tokens, i'd also recommend submitting a pr to https://github.com/dqbd/tiktokenizer/ because i don't think the special token ids are correct there (and a lot of the team uses tiktokenizer so it would be bad for incorrect tokens to be shown). [this is the tiktokenizer commit](https://github.com/dqbd/tiktokenizer/commit/071fb6e059d792aaaf2afbd8e11ece6c9d580840) which i believe may not be correct

File: src/openai_public.rs (hunk 1 above)
Comment (rishabhy) [reply src/openai_public.rs]: updated!

**Review by arvid220u — APPROVED** [review]
lgtm!

**Pull request merged**